openmls/src/binary_tree/array_representation/diff.rs

```diff
@@ -0,0 +1,602 @@
+    ) -> Result<usize, ABinaryTreeDiffError> {
+        // If the given leaf indices are identical, the shared subtree root is
+        // the index itself. Since the index of the leaf itself doesn't appear
+        // in the direct path, we can't return anything meaningful. This check
+        // also ensures that the tree is large enough such that the direct path
+        // is never empty, since if there is a second leaf index (that is within
+        // the bound of the tree), there is a non-leaf root node that is in the
+        // direct path of all leaves.
+        if leaf_index_1 == leaf_index_2 {
+            return Err(ABinaryTreeDiffError::SameLeafError);
+        } else if leaf_index_1 >= self.leaf_count() || leaf_index_2 >= self.leaf_count() {
+            return Err(ABinaryTreeDiffError::OutOfBounds);
+        }
+        let subtree_root_node_index =
+            lowest_common_ancestor(to_node_index(leaf_index_1), to_node_index(leaf_index_2));
+        let leaf_index_1_direct_path = direct_path(to_node_index(leaf_index_1), self.size())?;
+
+        leaf_index_1_direct_path
+            .iter()
+            .position(|&direct_path_node_index| direct_path_node_index == subtree_root_node_index)
+            // The shared subtree root has to be in the direct path of both nodes.
+            .ok_or(ABinaryTreeDiffError::LibraryError)
+    }
+
+    /// Returns `NodeReference` to the copath node of the `leaf_index_1` that is
+    /// in the direct path of `leaf_index_2`. Returns an error if both leaf
+    /// indices are identical or if one of the leaf indices does not correspond
+    /// to a leaf in the diff.
+    pub(crate) fn subtree_root_copath_node(
+        &'a self,
+        leaf_index_1: LeafIndex,
+        leaf_index_2: LeafIndex,
+    ) -> Result<NodeReference, ABinaryTreeDiffError> {
+        if leaf_index_1 == leaf_index_2 {
+            return Err(ABinaryTreeDiffError::SameLeafError);
+        } else if leaf_index_1 >= self.leaf_count() || leaf_index_2 >= self.leaf_count() {
+            return Err(ABinaryTreeDiffError::OutOfBounds);
+        }
+
+        // We want to return the position of the lowest common ancestor in the
+        // direct path of `leaf_index_1` (i.e. the sender_leaf_index).
+        let subtree_root_node_index =
+            lowest_common_ancestor(to_node_index(leaf_index_1), to_node_index(leaf_index_2));
+
+        // Figure out which child is the relevant copath node.
+        let copath_node_index = if leaf_index_2 < leaf_index_1 {
+            left(subtree_root_node_index)?
+        } else {
+            right(subtree_root_node_index, self.size())?
+        };
+
+        NodeReference::try_from_node_index(self, copath_node_index)
+    }
+
+    /// Returns a vector of `NodeReference`s, where the first reference is to
+    /// the root of the shared subtree of the two given leaf indices followed by
+    /// references to the nodes in its direct path. Returns an error if either
```

Comment:
Maybe worth clarifying that 'it' refers to the common ancestor and not to either leaf index.